File: BAEKJOON/2343 - 기타 레슨/기타 레슨.cs
using System;
using System.Linq;

class Program
{
    // 블루레이 개수를 계산하는 메서드
    static int CountBlurays(int[] lectures, int size) {
        int count = 1;  // 블루레이 개수 (최소 1부터 시작)
        int total = 0;  // 현재 블루레이에 담긴 강의 길이 합

        foreach (int lecture in lectures) {
            // 현재 블루레이에 강의를 추가했을 때 크기를 초과하는 경우
            if (total + lecture > size) {
                count++;  // 새로운 블루레이 필요
                total = lecture;  // 현재 강의를 새로운 블루레이에 담음
            } else {
                total += lecture;  // 크기를 초과하지 않으면 현재 블루레이에 추가
            }
        }
        return count;  // 사용된 블루레이 개수 반환
    }

    // 최소 블루레이 크기를 찾는 이분 탐색 메서드
    static int FindMinimumBluraySize(int n, int m, int[] lectures) {
        int left = lectures.Max();  // 가장 긴 강의 길이 (최소값)
        int right = lectures.Sum();  // 모든 강의 길이의 합 (최대값)
        int answer = right;  // 정답을 저장할 변수, 초기값은 최대값으로 설정

        // 이분 탐색 수행
        while (left <= right) {
            int mid = (left + right) / 2;  // 블루레이 크기의 중간값 설정
            if (CountBlurays(lectures, mid) <= m) {
                answer = mid;  // 블루레이 개수가 M개 이하이면 정답 갱신
                right = mid - 1;  // 더 작은 크기를 탐색
            } else {
                left = mid + 1;  // 블루레이 개수가 M개 초과하면 더 큰 크기를 탐색
            }
        }
        return answer;  // 최소 블루레이 크기 반환
    }

    static void Main() {
        string[] input = Console.ReadLine().Split(' ');
        int n = int.Parse(input[0]);  // 강의 개수
        int m = int.Parse(input[1]);  // 블루레이 개수
        int[] lectures = Console.ReadLine().Split(' ').Select(int.Parse).ToArray();  // 강의 길이 배열 입력

        Console.WriteLine(FindMinimumBluraySize(n, m, lectures));  // 최소 블루레이 크기 출력
    }
}

File: BAEKJOON/2343 - 기타 레슨/기타 레슨.py
def count_blurays(lectures, size):
    count = 1  # 사용된 블루레이 개수를 세기 위한 변수. 최소 1개의 블루레이는 필요하므로 1로 초기화
    total = 0  # 현재 블루레이에 담긴 강의 길이 합을 저장하는 변수

    for lecture in lectures:  # 강의 목록을 순회하면서 각 강의를 블루레이에 담음
        if (
            total + lecture > size
        ):  # 현재 블루레이에 강의를 추가했을 때 블루레이 크기를 초과하는 경우
            count += 1  # 새로운 블루레이가 필요하므로 개수 증가
            total = lecture  # 새 블루레이에 현재 강의를 담음 (누적 합 초기화)
        else:
            total += lecture  # 크기를 초과하지 않으면 현재 블루레이에 강의를 추가

    return count  # 사용된 블루레이의 총 개수 반환


def find_minimum_bluray_size(n, m, lectures):
    left = max(lectures)  # 블루레이 크기의 최소값은 가장 긴 강의의 길이
    right = sum(
        lectures
    )  # 블루레이 크기의 최대값은 모든 강의를 하나의 블루레이에 담았을 때의 총합
    answer = (
        right  # 가능한 최소 블루레이 크기를 저장하는 변수. 초기값은 최대값으로 설정
    )

    while left <= right:  # 이분 탐색 시작
        mid = (left + right) // 2  # 중간값을 블루레이 크기로 설정
        if (
            count_blurays(lectures, mid) <= m
        ):  # 중간값 크기에서 필요한 블루레이 개수가 M 이하인 경우
            answer = (
                mid  # 가능한 블루레이 크기 중 더 작은 값을 찾기 위해 현재 값을 저장
            )
            right = mid - 1  # 블루레이 크기를 줄이기 위해 탐색 범위를 왼쪽으로 이동
        else:  # 중간값에서 필요한 블루레이 개수가 M보다 큰 경우
            left = mid + 1  # 블루레이 크기를 늘리기 위해 탐색 범위를 오른쪽으로 이동

    return answer  # 최종적으로 찾은 최소 블루레이 크기 반환


n, m = map(int, input().split())  # 강의의 개수 N과 블루레이 개수 M 입력
lectures = list(map(int, input().split()))  # 각 강의의 길이를 입력받아 리스트로 저장

print(find_minimum_bluray_size(n, m, lectures))  # 최소 블루레이 크기를 계산하고 출력
